File: acs-integration-tests/src/test/java/org/eclipse/keti/test/utils/PrivilegeHelper.java
/*******************************************************************************
 * Copyright 2017 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *******************************************************************************/

package org.eclipse.keti.test.utils;

import static org.eclipse.keti.test.utils.ACSTestUtil.ACS_VERSION;

import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.security.oauth2.client.OAuth2RestTemplate;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import org.eclipse.keti.acs.model.Attribute;
import org.eclipse.keti.acs.rest.BaseResource;
import org.eclipse.keti.acs.rest.BaseSubject;

@Component
public class PrivilegeHelper {

    private static final String ALTERNATE_ATTRIBUTE_VALUE = "sanfrancisco";
    private static final String DEFAULT_ATTRIBUTE_VALUE = "sanramon";
    private static final String DEFAULT_ATTRIBUTE_NAME = "site";

    private static final String DEFAULT_ORG_ATTRIBUTE_NAME = "org";
    private static final String DEFAULT_ORG_ATTRIBUTE_VALUE = "alliance";
    private static final String ALTERNATE_ORG_ATTRIBUTE_VALUE = "syndicate";

    public static final String ACS_SUBJECT_API_PATH = ACS_VERSION + "/subject/";
    public static final String ACS_RESOURCE_API_PATH = ACS_VERSION + "/resource/";

    public static final String DEFAULT_SUBJECT_ID = "any";
    public static final String DEFAULT_SUBJECT_IDENTIFIER = "any";
    public static final String DEFAULT_RESOURCE_IDENTIFIER = "any";
    public static final String DEFAULT_ATTRIBUTE_ISSUER = "https://acs.attributes.int";

    @Autowired
    private ZoneFactory zoneFactory;

    public BaseSubject putSubject(final OAuth2RestTemplate acsTemplate, final BaseSubject subject,
            final String endpoint, final HttpHeaders headers, final Attribute... attributes)
            throws UnsupportedEncodingException {

        subject.setAttributes(new HashSet<>(Arrays.asList(attributes)));
        URI subjectUri = URI
                .create(endpoint + ACS_SUBJECT_API_PATH + URLEncoder.encode(subject.getSubjectIdentifier(), "UTF-8"));
        acsTemplate.put(subjectUri, new HttpEntity<>(subject, headers));
        return subject;
    }

    public void putSubject(final OAuth2RestTemplate restTemplate, final String subjectIdentifier,
            final HttpHeaders headers, final Attribute... attributes) throws UnsupportedEncodingException {

        BaseSubject subject = new BaseSubject(subjectIdentifier);
        // no header needed, because it uses zone specific url
        putSubject(restTemplate, subject, this.zoneFactory.getAcsBaseURL(), headers, attributes);
    }

    public ResponseEntity<Object> postMultipleSubjects(final OAuth2RestTemplate acsTemplate, final String endpoint,
            final HttpHeaders headers, final BaseSubject... subjects) {
        Attribute site = getDefaultAttribute();
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(site);

        List<BaseSubject> subjectsArray = new ArrayList<>();
        for (BaseSubject s : subjects) {
            s.setAttributes(attributes);
            subjectsArray.add(s);
        }
        URI subjectUri = URI.create(endpoint + ACS_SUBJECT_API_PATH);

        ResponseEntity<Object> responseEntity = acsTemplate.postForEntity(subjectUri,
                new HttpEntity<>(subjectsArray, headers), Object.class);

        return responseEntity;
    }

    public ResponseEntity<Object> postSubjectsWithDefaultAttributes(final OAuth2RestTemplate acsTemplate,
            final String endpoint, final HttpHeaders headers, final BaseSubject... subjects) {
        Attribute site = getDefaultAttribute();
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(site);

        for (BaseSubject s : subjects) {
            s.setAttributes(attributes);
        }
        return postSubjects(acsTemplate, endpoint, headers, subjects);
    }

    public ResponseEntity<Object> postSubjects(final OAuth2RestTemplate acsTemplate, final String endpoint,
            final HttpHeaders headers, final BaseSubject... subjects) {
        List<BaseSubject> subjectsArray = new ArrayList<>();
        for (BaseSubject s : subjects) {
            subjectsArray.add(s);
        }
        URI subjectUri = URI.create(endpoint + ACS_SUBJECT_API_PATH);
        ResponseEntity<Object> responseEntity = acsTemplate.postForEntity(subjectUri,
                new HttpEntity<>(subjectsArray, headers), Object.class);
        return responseEntity;
    }

    public BaseResource putResource(final OAuth2RestTemplate acsTemplate, final BaseResource resource,
            final String endpoint, final HttpHeaders headers, final Attribute... attributes) throws Exception {

        resource.setAttributes(new HashSet<>(Arrays.asList(attributes)));

        String value = URLEncoder.encode(resource.getResourceIdentifier(), "UTF-8");

        URI uri = new URI(endpoint + ACS_RESOURCE_API_PATH + value);
        acsTemplate.put(uri, new HttpEntity<>(resource, headers));
        return resource;
    }

    public ResponseEntity<Object> postResources(final OAuth2RestTemplate acsTemplate, final String endpoint,
            final BaseResource... resources) {

        Attribute site = getDefaultAttribute();
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(site);

        List<BaseResource> resourcesArray = new ArrayList<>();
        for (BaseResource r : resources) {
            r.setAttributes(attributes);
            resourcesArray.add(r);
        }
        URI resourceUri = URI.create(endpoint + ACS_RESOURCE_API_PATH);

        ResponseEntity<Object> responseEntity = acsTemplate.postForEntity(resourceUri, resourcesArray, Object.class);

        return responseEntity;
    }

    public ResponseEntity<Object> postResourcesWithDefaultSiteAttribute(final OAuth2RestTemplate acsTemplate,
            final String endpoint, final HttpHeaders headers, final BaseResource... resources) {

        Attribute site = getDefaultAttribute();
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(site);

        for (BaseResource r : resources) {
            r.setAttributes(attributes);
        }
        return postResources(acsTemplate, endpoint, headers, resources);
    }

    public ResponseEntity<Object> postResources(final OAuth2RestTemplate acsTemplate, final String endpoint,
            final HttpHeaders headers, final BaseResource... resources) {

        List<BaseResource> resourcesArray = new ArrayList<>();
        for (BaseResource r : resources) {
            resourcesArray.add(r);
        }
        URI resourceUri = URI.create(endpoint + ACS_RESOURCE_API_PATH);

        ResponseEntity<Object> responseEntity = acsTemplate.postForEntity(resourceUri,
                new HttpEntity<>(resourcesArray, headers), Object.class);

        return responseEntity;
    }

    public void deleteSubject(final RestTemplate restTemplate, final String acsUrl, final String subjectId)
            throws Exception {
        deleteSubject(restTemplate, acsUrl, subjectId, null);
    }

    public void deleteSubject(final RestTemplate restTemplate, final String acsUrl, final String subjectId,
            final HttpHeaders headers) throws Exception {
        if (subjectId != null) {
            URI subjectUri = URI.create(acsUrl + ACS_SUBJECT_API_PATH + URLEncoder.encode(subjectId, "UTF-8"));
            restTemplate.exchange(subjectUri, HttpMethod.DELETE, new HttpEntity<>(headers), String.class);
        }
    }

    public void deleteResource(final RestTemplate restTemplate, final String acsUrl, final String resourceId)
            throws Exception {
        deleteResource(restTemplate, acsUrl, resourceId, null);
    }

    public void deleteResource(final RestTemplate restTemplate, final String acsUrl, final String resourceId,
            final HttpHeaders headers) throws Exception {
        if (resourceId != null) {
            URI resourceUri = URI.create(acsUrl + ACS_RESOURCE_API_PATH + URLEncoder.encode(resourceId, "UTF-8"));
            restTemplate.exchange(resourceUri, HttpMethod.DELETE, new HttpEntity<>(headers), String.class);
        }
    }

    public BaseSubject createSubject(final String subjectIdentifier) {
        BaseSubject subject = new BaseSubject();
        subject.setSubjectIdentifier(subjectIdentifier);
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(getDefaultAttribute());
        subject.setAttributes(attributes);
        return subject;
    }

    public BaseResource createResource(final String resourceIdentifier) {
        BaseResource resource = new BaseResource();
        resource.setResourceIdentifier(resourceIdentifier);
        Set<Attribute> attributes = new HashSet<>();
        attributes.add(getDefaultAttribute());
        resource.setAttributes(attributes);
        return resource;
    }

    public Attribute getDefaultAttribute() {
        Attribute site = new Attribute();
        site.setIssuer(DEFAULT_ATTRIBUTE_ISSUER);
        site.setName(DEFAULT_ATTRIBUTE_NAME);
        site.setValue(DEFAULT_ATTRIBUTE_VALUE);
        return site;
    }

    public Attribute getAlternateAttribute() {
        Attribute site = new Attribute();
        site.setIssuer(DEFAULT_ATTRIBUTE_ISSUER);
        site.setName(DEFAULT_ATTRIBUTE_NAME);
        site.setValue(ALTERNATE_ATTRIBUTE_VALUE);
        return site;
    }

    public Attribute getDefaultOrgAttribute() {
        Attribute site = new Attribute();
        site.setIssuer(DEFAULT_ATTRIBUTE_ISSUER);
        site.setName(DEFAULT_ORG_ATTRIBUTE_NAME);
        site.setValue(DEFAULT_ORG_ATTRIBUTE_VALUE);
        return site;
    }

    public Attribute getAlternateOrgAttribute() {
        Attribute site = new Attribute();
        site.setIssuer(DEFAULT_ATTRIBUTE_ISSUER);
        site.setName(DEFAULT_ORG_ATTRIBUTE_NAME);
        site.setValue(ALTERNATE_ORG_ATTRIBUTE_VALUE);
        return site;
    }

    public void deleteSubjects(final RestTemplate restTemplate, final String acsUrl, final HttpHeaders headers)
            throws Exception {
        BaseSubject[] subjects = listSubjects(restTemplate, acsUrl, headers);
        for (BaseSubject subject : subjects) {
            deleteSubject(restTemplate, acsUrl, subject.getSubjectIdentifier(), headers);
        }
    }

    public BaseSubject[] listSubjects(final RestTemplate restTemplate, final String acsUrl, final HttpHeaders headers) {
        URI uri = URI.create(acsUrl + ACS_SUBJECT_API_PATH);
        ResponseEntity<BaseSubject[]> response = restTemplate.exchange(uri, HttpMethod.GET, new HttpEntity<>(headers),
                BaseSubject[].class);
        return response.getBody();
    }

    public void deleteResources(final RestTemplate restTemplate, final String acsUrl, final HttpHeaders headers)
            throws Exception {
        BaseResource[] resources = listResources(restTemplate, acsUrl, headers);
        for (BaseResource resource : resources) {
            deleteResource(restTemplate, acsUrl, resource.getResourceIdentifier(), headers);
        }
    }

    public BaseResource[] listResources(final RestTemplate restTemplate, final String acsEndpoint,
            final HttpHeaders headers) {
        URI uri = URI.create(acsEndpoint + ACS_RESOURCE_API_PATH);
        ResponseEntity<BaseResource[]> response = restTemplate.exchange(uri, HttpMethod.GET, new HttpEntity<>(headers),
                BaseResource[].class);
        return response.getBody();
    }
}

File: acs-integration-tests/src/test/java/org/eclipse/keti/test/utils/PrivilegeHelper.kt
/*******************************************************************************
 * Copyright 2018 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.keti.test.utils

import org.eclipse.keti.acs.model.Attribute
import org.eclipse.keti.acs.rest.BaseResource
import org.eclipse.keti.acs.rest.BaseSubject
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.security.oauth2.client.OAuth2RestTemplate
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
import java.io.UnsupportedEncodingException
import java.net.URI
import java.net.URLEncoder
import java.util.ArrayList
import java.util.Arrays
import java.util.HashSet

private const val ALTERNATE_ATTRIBUTE_VALUE = "sanfrancisco"
private const val DEFAULT_ATTRIBUTE_VALUE = "sanramon"
private const val DEFAULT_ATTRIBUTE_NAME = "site"
private const val DEFAULT_ORG_ATTRIBUTE_NAME = "org"
private const val DEFAULT_ORG_ATTRIBUTE_VALUE = "alliance"
private const val ALTERNATE_ORG_ATTRIBUTE_VALUE = "syndicate"

const val ACS_SUBJECT_API_PATH = "$ACS_VERSION/subject/"
const val ACS_RESOURCE_API_PATH = "$ACS_VERSION/resource/"
const val DEFAULT_SUBJECT_ID = "any"
const val DEFAULT_SUBJECT_IDENTIFIER = "any"
const val DEFAULT_RESOURCE_IDENTIFIER = "any"
const val DEFAULT_ATTRIBUTE_ISSUER = "https://acs.attributes.int"

@Component
class PrivilegeHelper {

    @Autowired
    private lateinit var zoneFactory: ZoneFactory

    val defaultAttribute: Attribute
        get() {
            val site = Attribute()
            site.issuer = DEFAULT_ATTRIBUTE_ISSUER
            site.name = DEFAULT_ATTRIBUTE_NAME
            site.value = DEFAULT_ATTRIBUTE_VALUE
            return site
        }

    val alternateAttribute: Attribute
        get() {
            val site = Attribute()
            site.issuer = DEFAULT_ATTRIBUTE_ISSUER
            site.name = DEFAULT_ATTRIBUTE_NAME
            site.value = ALTERNATE_ATTRIBUTE_VALUE
            return site
        }

    val defaultOrgAttribute: Attribute
        get() {
            val site = Attribute()
            site.issuer = DEFAULT_ATTRIBUTE_ISSUER
            site.name = DEFAULT_ORG_ATTRIBUTE_NAME
            site.value = DEFAULT_ORG_ATTRIBUTE_VALUE
            return site
        }

    val alternateOrgAttribute: Attribute
        get() {
            val site = Attribute()
            site.issuer = DEFAULT_ATTRIBUTE_ISSUER
            site.name = DEFAULT_ORG_ATTRIBUTE_NAME
            site.value = ALTERNATE_ORG_ATTRIBUTE_VALUE
            return site
        }

    @Throws(UnsupportedEncodingException::class)
    fun putSubject(
        acsTemplate: OAuth2RestTemplate,
        subject: BaseSubject,
        endpoint: String?,
        headers: HttpHeaders,
        vararg attributes: Attribute
    ): BaseSubject {

        subject.attributes = HashSet(Arrays.asList(*attributes))
        val subjectUri = URI
            .create(endpoint + ACS_SUBJECT_API_PATH + URLEncoder.encode(subject.subjectIdentifier!!, "UTF-8"))
        acsTemplate.put(subjectUri, HttpEntity(subject, headers))
        return subject
    }

    @Throws(UnsupportedEncodingException::class)
    fun putSubject(
        restTemplate: OAuth2RestTemplate,
        subjectIdentifier: String,
        headers: HttpHeaders,
        vararg attributes: Attribute
    ) {

        val subject = BaseSubject(subjectIdentifier)
        // no header needed, because it uses zone specific url
        putSubject(restTemplate, subject, this.zoneFactory.acsBaseURL, headers, *attributes)
    }

    fun postMultipleSubjects(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        headers: HttpHeaders,
        vararg subjects: BaseSubject
    ): ResponseEntity<Any> {
        val site = defaultAttribute
        val attributes = HashSet<Attribute>()
        attributes.add(site)

        val subjectsArray = ArrayList<BaseSubject>()
        for (s in subjects) {
            s.attributes = attributes
            subjectsArray.add(s)
        }
        val subjectUri = URI.create(endpoint + ACS_SUBJECT_API_PATH)

        return acsTemplate.postForEntity(
            subjectUri,
            HttpEntity<List<BaseSubject>>(subjectsArray, headers), Any::class.java
        )
    }

    fun postSubjectsWithDefaultAttributes(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        headers: HttpHeaders,
        vararg subjects: BaseSubject
    ): ResponseEntity<Any> {
        val site = defaultAttribute
        val attributes = HashSet<Attribute>()
        attributes.add(site)

        for (s in subjects) {
            s.attributes = attributes
        }
        return postSubjects(acsTemplate, endpoint, headers, *subjects)
    }

    fun postSubjects(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        headers: HttpHeaders,
        vararg subjects: BaseSubject
    ): ResponseEntity<Any> {
        val subjectsArray = ArrayList<BaseSubject>()
        for (s in subjects) {
            subjectsArray.add(s)
        }
        val subjectUri = URI.create(endpoint + ACS_SUBJECT_API_PATH)
        return acsTemplate.postForEntity(
            subjectUri,
            HttpEntity<List<BaseSubject>>(subjectsArray, headers), Any::class.java
        )
    }

    @Throws(Exception::class)
    fun putResource(
        acsTemplate: OAuth2RestTemplate,
        resource: BaseResource,
        endpoint: String,
        headers: HttpHeaders,
        vararg attributes: Attribute
    ): BaseResource {

        resource.attributes = HashSet(Arrays.asList(*attributes))

        val value = URLEncoder.encode(resource.resourceIdentifier!!, "UTF-8")

        val uri = URI(endpoint + ACS_RESOURCE_API_PATH + value)
        acsTemplate.put(uri, HttpEntity(resource, headers))
        return resource
    }

    fun postResources(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        vararg resources: BaseResource
    ): ResponseEntity<Any> {

        val site = defaultAttribute
        val attributes = HashSet<Attribute>()
        attributes.add(site)

        val resourcesArray = ArrayList<BaseResource>()
        for (r in resources) {
            r.attributes = attributes
            resourcesArray.add(r)
        }
        val resourceUri = URI.create(endpoint + ACS_RESOURCE_API_PATH)

        return acsTemplate.postForEntity(resourceUri, resourcesArray, Any::class.java)
    }

    fun postResourcesWithDefaultSiteAttribute(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        headers: HttpHeaders,
        vararg resources: BaseResource
    ): ResponseEntity<Any> {

        val site = defaultAttribute
        val attributes = HashSet<Attribute>()
        attributes.add(site)

        for (r in resources) {
            r.attributes = attributes
        }
        return postResources(acsTemplate, endpoint, headers, *resources)
    }

    fun postResources(
        acsTemplate: OAuth2RestTemplate,
        endpoint: String,
        headers: HttpHeaders,
        vararg resources: BaseResource
    ): ResponseEntity<Any> {

        val resourcesArray = ArrayList<BaseResource>()
        for (r in resources) {
            resourcesArray.add(r)
        }
        val resourceUri = URI.create(endpoint + ACS_RESOURCE_API_PATH)

        return acsTemplate.postForEntity(
            resourceUri,
            HttpEntity<List<BaseResource>>(resourcesArray, headers), Any::class.java
        )
    }

    @Throws(Exception::class)
    @JvmOverloads
    fun deleteSubject(
        restTemplate: RestTemplate,
        acsUrl: String,
        subjectId: String?,
        headers: HttpHeaders? = null
    ) {
        if (subjectId != null) {
            val subjectUri = URI.create(acsUrl + ACS_SUBJECT_API_PATH + URLEncoder.encode(subjectId, "UTF-8"))
            restTemplate.exchange(subjectUri, HttpMethod.DELETE, HttpEntity<Any>(headers), String::class.java)
        }
    }

    @Throws(Exception::class)
    @JvmOverloads
    fun deleteResource(
        restTemplate: RestTemplate,
        acsUrl: String,
        resourceId: String?,
        headers: HttpHeaders? = null
    ) {
        if (resourceId != null) {
            val resourceUri = URI.create(acsUrl + ACS_RESOURCE_API_PATH + URLEncoder.encode(resourceId, "UTF-8"))
            restTemplate.exchange(resourceUri, HttpMethod.DELETE, HttpEntity<Any>(headers), String::class.java)
        }
    }

    fun createSubject(subjectIdentifier: String): BaseSubject {
        val subject = BaseSubject()
        subject.subjectIdentifier = subjectIdentifier
        val attributes = HashSet<Attribute>()
        attributes.add(defaultAttribute)
        subject.attributes = attributes
        return subject
    }

    fun createResource(resourceIdentifier: String): BaseResource {
        val resource = BaseResource()
        resource.resourceIdentifier = resourceIdentifier
        val attributes = HashSet<Attribute>()
        attributes.add(defaultAttribute)
        resource.attributes = attributes
        return resource
    }

    @Throws(Exception::class)
    fun deleteSubjects(
        restTemplate: RestTemplate,
        acsUrl: String,
        headers: HttpHeaders
    ) {
        val subjects = listSubjects(restTemplate, acsUrl, headers)
        for (subject in subjects) {
            deleteSubject(restTemplate, acsUrl, subject.subjectIdentifier, headers)
        }
    }

    fun listSubjects(
        restTemplate: RestTemplate,
        acsUrl: String,
        headers: HttpHeaders
    ): Array<BaseSubject> {
        val uri = URI.create(acsUrl + ACS_SUBJECT_API_PATH)
        val response = restTemplate.exchange(
            uri, HttpMethod.GET, HttpEntity<Any>(headers),
            Array<BaseSubject>::class.java
        )
        return response.body
    }

    @Throws(Exception::class)
    fun deleteResources(
        restTemplate: RestTemplate,
        acsUrl: String,
        headers: HttpHeaders
    ) {
        val resources = listResources(restTemplate, acsUrl, headers)
        for (resource in resources) {
            deleteResource(restTemplate, acsUrl, resource.resourceIdentifier, headers)
        }
    }

    fun listResources(
        restTemplate: RestTemplate,
        acsEndpoint: String,
        headers: HttpHeaders
    ): Array<BaseResource> {
        val uri = URI.create(acsEndpoint + ACS_RESOURCE_API_PATH)
        val response = restTemplate.exchange(
            uri, HttpMethod.GET, HttpEntity<Any>(headers),
            Array<BaseResource>::class.java
        )
        return response.body
    }
}
